qupulse/hardware/awgs/base.py

```diff
@@ -1,166 +1,202 @@
-"""This module defines the common interface for arbitrary waveform generators.
+import warnings
+from abc import ABC, abstractmethod
+from copy import copy
+from typing import TypeVar, Generic, List, Iterable, Dict, Callable
 
-Classes:
-    - AWG: Common AWG interface.
-    - DummyAWG: A software stub implementation of the AWG interface.
-    - ProgramOverwriteException
-    - OutOfWaveformMemoryException
-"""
 
-from abc import abstractmethod
-from typing import Set, Tuple, Callable, Optional
+class BaseFeature(ABC):
```

Comment:
Does it make sense to make BaseFeature a metaclass for better static analysis?

Reply:
What do you mean with "static analysis", @terrorfisch ? Do you mean the syntax checking, that is performed by the IDE (or mypy, for example)? If yes, then I didn't find a way that really works. If you have an idea to solve this, feel free to show us :)

Reply:
Yes, basically for IDE support. It works with warnings for abstractmethods but maybe because this is hard baked into the analyzers. I'll test my idea with jedi which is the analyzer Spyder uses afaik.